### PR TITLE
Add more output matrices, FIX HORRIFIC BUG WITH TIMING

### PR DESCRIPTION
**Pull request by helenol**
Total times used to wrap around after 60.0 seconds.
To 0.
But luckily only in the print output, so that average times, etc. are not affected.

Other than that, added A not-inverse and M pseudo-inverse output for better constraint packing and repacking, and added a test for it.

@SebastianInd FYI

### CODE REVIEW

File: mav_trajectory_generation/test/test_polynomial_optimization.cpp [hunk 1]
```diff
@@ -775,6 +773,69 @@ TEST(MavTrajectoryGeneration, 2_vertices_rand) {
   }
 }
 
+// Test unpacking and repacking constraints between [d_f; d_p] and p.
+// Test 2 vertices setup
```
Comment (rikba) [inline mav_trajectory_generation/test/test_polynomial_optimization.cpp]: Remove.

File: mav_trajectory_generation/test/test_polynomial_optimization.cpp [hunk 2]
```diff
@@ -775,6 +773,69 @@ TEST(MavTrajectoryGeneration, 2_vertices_rand) {
...
+
+    // For each dimension... We can test that [df;dp] -> p -> [df;dp].
+    for (int i = 0; i < 3; ++i) {
+      Eigen::VectorXd d_all_ordered(fixed_constraints[i].size() +
+                                    free_constraints[i].size());
+      d_all_ordered << fixed_constraints[i], free_constraints[i];
+      Eigen::VectorXd p = A_inv * M * d_all_ordered;
+      Eigen::VectorXd d_unordered = A * p;
+      Eigen::VectorXd d_reordered = M_pinv * d_unordered;
+      EXPECT_TRUE(EIGEN_MATRIX_NEAR(d_all_ordered, d_reordered, 1e-6));
```
Comment (rikba) [inline mav_trajectory_generation/test/test_polynomial_optimization.cpp]: Just to make sure. This check makes sure `getM`, `getAInverse`, `getA`, `getMpinv` are correct.

File: mav_trajectory_generation/test/test_polynomial_optimization.cpp [hunk 3]
```diff
@@ -775,6 +773,69 @@ TEST(MavTrajectoryGeneration, 2_vertices_rand) {
...
+
+      // Now also check for each segment.
+      for (size_t j = 0; j < segments.size(); ++j) {
+        Eigen::VectorXd p_seg = segments[j][i].getCoefficients(0);
+        EXPECT_TRUE(EIGEN_MATRIX_NEAR(p_seg, p.segment<N>(j * N), 1e-6));
```
Comment (rikba) [inline mav_trajectory_generation/test/test_polynomial_optimization.cpp]: This is an extra check for `A_inv` and `M` being correct.

File: mav_trajectory_generation/include/mav_trajectory_generation/impl/polynomial_optimization_linear_impl.h [hunk 4]
```diff
@@ -541,6 +542,35 @@ void PolynomialOptimization<_N>::getR(Eigen::MatrixXd* R) const {
...
+}
+
+template <int _N>
+void PolynomialOptimization<_N>::getMpinv(Eigen::MatrixXd* M_pinv) const {
+  CHECK_NOTNULL(M_pinv);
+
+  // Pseudoinverse implementation by @SebastianInd.
+  *M_pinv = constraint_reordering_.transpose();
+  for (int M_row = 0; M_row < M_pinv->rows(); M_row++) {
+    M_pinv->row(M_row) = M_pinv->row(M_row) / M_pinv->row(M_row).sum();
```
Comment (rikba) [inline mav_trajectory_generation/include/mav_trajectory_generation/impl/polynomial_optimization_linear_impl.h]: Is this based on some special form `M` has? Looks super short. Nice!

File: mav_trajectory_generation/test/test_polynomial_optimization.cpp (hunk 2 above)
Comment (helenol) [inline mav_trajectory_generation/test/test_polynomial_optimization.cpp]: Yes. :) This checks that converting back and forth between the polynomial coefficients and the packed/unpacked constraints works correctly. The content of A_inv and M are are checked when repacking the ps from d fixed and d parameters and then comparing it with the ps stored in segments. A and M pinv are checked by taking the ps are unpacking them again into d_unoreded and d_reordered.

I check the content of A specifically against the MATLAB implementation in some tests in local planning, but really these two matrices are only ever used together to switch between parameter orderings.

File: mav_trajectory_generation/test/test_polynomial_optimization.cpp (hunk 3 above)
Comment (helenol) [inline mav_trajectory_generation/test/test_polynomial_optimization.cpp]: Si

File: mav_trajectory_generation/include/mav_trajectory_generation/impl/polynomial_optimization_linear_impl.h (hunk 4 above)
Comment (helenol) [inline mav_trajectory_generation/include/mav_trajectory_generation/impl/polynomial_optimization_linear_impl.h]: All Sebastian's work. ;) But yes. M is a matrix of only 1s and 0s for reordering, so to get the pseudoinverse just need to make sure all rows sum up to 1 in the final matrix.

**Pull request merged**